core/src/main/java/org/infinispan/protostream/impl/TagWriterImpl.java

```diff
@@ -794,6 +819,51 @@ void writeBytes(ByteBuffer value) throws IOException {
          }
       }
 
+      @Override
+      void writeUTF8Field(int number, String s) throws IOException {
+         if (!(out instanceof LazyByteArrayOutputStream os)) {
+            super.writeUTF8Field(number, s);
+            return;
+         }
+
+         int strlen = s.length();
+         int tag = WireType.makeTag(number, WireType.WIRETYPE_LENGTH_DELIMITED);
+         int varIntTagLen = varIntSize(tag);
+         int varIntBytesLen = varIntSize(strlen);
+         int startPos = os.getPosition();
+
+         // First optimize for 1 - 127 case
+         os.ensureCapacity(startPos + varIntTagLen + varIntBytesLen + strlen);
+         byte[] buf = os.getRawBuffer();
+         writeVarInt32Direct(tag, startPos, buf);
+         // Note this will be overwritten if not all 1 - 127 characters below
+         writeVarInt32Direct(strlen, startPos + varIntTagLen, buf);
+
+         int localPos = startPos + varIntTagLen + varIntBytesLen;
+
+         int c;
+         int i;
+         for (i = 0; i < strlen; i++) {
+            c = s.charAt(i);
+            if (c > 127) break;
+
+            buf[localPos++] = (byte) c;
+         }
+
+         os.setPosition(localPos);
+         // All single byte characters
+         if (i == strlen)
+            return;
+
+         byte[] utf8buffer = s.getBytes(StandardCharsets.UTF_8);
+         varIntBytesLen = varIntSize(utf8buffer.length);
+         os.ensureCapacity(startPos + varIntTagLen + varIntBytesLen + utf8buffer.length);
+         buf = os.getRawBuffer();
+         writeVarInt32Direct(utf8buffer.length, startPos + varIntTagLen, buf);
+         System.arraycopy(utf8buffer, 0, buf, startPos + varIntTagLen + varIntBytesLen, utf8buffer.length);
```

Comment:
Shouldn't this be `out.writeBytes(uft8Buffer, 0, buf)`? Then the setPosition call wouldn't be required either.

Reply:
I think it's similar, as calling `os.write()` will result in `ensureCapacity` being called internally again, which we know is sufficient as we have already expanded the buf (if required).